mujoco_warp/_src/io.py

```diff
@@ -1649,3 +1653,202 @@ def get_data_into(
+  nf_out[worldid] = 0
+  nl_out[worldid] = 0
+  nefc_out[worldid] = 0
+  if worldid == 0:
+    nsolving_out[0] = nworld_in
+  time_out[worldid] = 0.0
+  energy_out[worldid] = wp.vec2(0.0, 0.0)
+
+
+@wp.kernel
+def _zero_nv(
+  # Data out:
+  qvel_out: wp.array2d(dtype=float),
+  qacc_warmstart_out: wp.array2d(dtype=float),
+  qfrc_applied_out: wp.array2d(dtype=float),
+  qacc_out: wp.array2d(dtype=float),
+):
+  worldid, dofid = wp.tid()
+  qvel_out[worldid, dofid] = 0.0
+  qacc_warmstart_out[worldid, dofid] = 0.0
+  qfrc_applied_out[worldid, dofid] = 0.0
+  qacc_out[worldid, dofid] = 0.0
+
+
+@wp.kernel
+def _zero_nu(
+  # Model:
+  na: int,
+  # Data out:
+  act_out: wp.array2d(dtype=float),
+  ctrl_out: wp.array2d(dtype=float),
+  act_dot_out: wp.array2d(dtype=float),
+):
+  worldid, actid = wp.tid()
+  ctrl_out[worldid, actid] = 0.0
+  if actid < na:
+    act_out[worldid, actid] = 0.0
+    act_dot_out[worldid, actid] = 0.0
+
+
+@wp.kernel
+def _set_mocap(
+  # Model:
+  body_mocapid: wp.array(dtype=int),
+  body_pos: wp.array2d(dtype=wp.vec3),
+  body_quat: wp.array2d(dtype=wp.quat),
+  # Data out:
+  mocap_pos_out: wp.array2d(dtype=wp.vec3),
+  mocap_quat_out: wp.array2d(dtype=wp.quat),
+):
+  worldid, bodyid = wp.tid()
+
+  mocapid = body_mocapid[bodyid]
+
+  if mocapid >= 0:
+    mocap_pos_out[worldid, mocapid] = body_pos[worldid, bodyid]
+    mocap_quat_out[worldid, mocapid] = body_quat[worldid, bodyid]
+
+
+@wp.kernel
+def _zero_contact(
+  # Data in:
+  ncon_in: wp.array(dtype=int),
+  # In:
+  nefcaddress: int,
+  # Data out:
+  contact_dist_out: wp.array(dtype=float),
+  contact_pos_out: wp.array(dtype=wp.vec3),
+  contact_frame_out: wp.array(dtype=wp.mat33),
+  contact_includemargin_out: wp.array(dtype=float),
+  contact_friction_out: wp.array(dtype=types.vec5),
+  contact_solref_out: wp.array(dtype=wp.vec2),
+  contact_solreffriction_out: wp.array(dtype=wp.vec2),
+  contact_solimp_out: wp.array(dtype=types.vec5),
+  contact_dim_out: wp.array(dtype=int),
+  contact_geom_out: wp.array(dtype=wp.vec2i),
+  contact_efc_address_out: wp.array2d(dtype=int),
+  contact_worldid_out: wp.array(dtype=int),
+):
+  conid = wp.tid()
+
+  if conid >= ncon_in[0]:
+    return
+
+  contact_dist_out[conid] = 0.0
+  contact_pos_out[conid] = wp.vec3(0.0)
+  contact_frame_out[conid] = wp.mat33(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0)
+  contact_includemargin_out[conid] = 0.0
+  contact_friction_out[conid] = types.vec5(0.0, 0.0, 0.0, 0.0, 0.0)
+  contact_solref_out[conid] = wp.vec2(0.0, 0.0)
+  contact_solreffriction_out[conid] = wp.vec2(0.0, 0.0)
+  contact_solimp_out[conid] = types.vec5(0.0, 0.0, 0.0, 0.0, 0.0)
+  contact_dim_out[conid] = 0
+  contact_geom_out[conid] = wp.vec2i(0, 0)
+  for i in range(nefcaddress):
+    contact_efc_address_out[conid, i] = 0
+  contact_worldid_out[conid] = 0
+
+
+def reset_data(m: types.Model, d: types.Data):
+  """Clear data, set defaults."""
+  # copy qpos0 from model
+  wp.copy(d.qpos, m.qpos0)
```

Comment:
i guess because of our stride 0 tricks this works just fine even with different `nworld`?

it probably does... but i would be a little more comfortable if the `test_reset_data` function below created test fixutres with `nworld` > 1 just to be sure

Reply:
updated to set `qpos` in the `_set_nworld` kernel (`qpos[worldid, i] = qpos0[worldid, i]`) and `test_reset_data` now sets `nworld=2` for testing.